doc: add more examples to core_section

Exirel opened the pull request:
### Description

I'm tired of user asking "how to provide multiple values?" for list, and I think our users are probably tired of asking, too! So I tried to add more examples whenever possible.

I'm sure I've miss a lot, and @dgw will probably have to fight my Nemesi**S**.

Oh well.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

dgw commented on sopel/config/core_section.py:
```diff
@@ -152,33 +189,72 @@ class CoreSection(StaticSection):
+
+    .. important::
+
+        When you edit the config file manually, make sure to escape each line
+        starting with a ``#`` with double quotes (as shown in the example
+        above), otherwise Python will ignore them, thinking they are comments.
```
I tried to improve this, with mixed results.

```suggestion
        If you edit the config file manually, make sure to wrap each line
        starting with a ``#`` in double quotes, as shown in the example above.
        An unquoted ``#`` denotes a comment, which will be ignored by Sopel's
        configuration parser.
```

Would a link to [the ConfigParser syntax docs](https://docs.python.org/3/library/configparser.html#supported-ini-file-structure) be overkill here?

Exirel replied:
No that's a good idea. I'll add this suggestion and I'll try not to forget about the link!

dgw commented on sopel/config/core_section.py:
```diff
@@ -152,33 +189,72 @@ class CoreSection(StaticSection):
-    Each line is a message that will be sent to the server once connected.
-    Example::
+    Each line is a message that will be sent to the server once connected,
+    in that order:
```
```suggestion
    in the order they are defined:
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -302,45 +390,108 @@ class CoreSection(StaticSection):
+
     Used when plugins format times with :func:`sopel.tools.time.format_time`.
+
+    Example with the British time::
```
```suggestion
    For example, to make Sopel fall back on British time::
```

(idk, still kinda messy)

dgw commented on sopel/config/core_section.py:
```diff
@@ -302,45 +390,108 @@ class CoreSection(StaticSection):
+            xkcd
+            help
+
+    In that case, only ``url``, ``xkcd``, and the ``help`` plugins will be
```
```suggestion
    In that case, only the ``url``, ``xkcd``, and ``help`` plugins will be
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -389,57 +564,86 @@ def homedir(self):
 
     :default: ``chat.freenode.net``
 
-    **Required.**
+    **Required**:
+
+    .. code-block:: ini
+
+        host = chat.freenode.net
+
     """
 
     host_blocks = ListAttribute('host_blocks')
     """A list of hostnames which Sopel should ignore.
 
-    Messages from any user whose connection hostname matches one of these values
-    will be ignored. :ref:`Regular expression syntax <re-syntax>` is supported.
+    Messages from any user whose connection hostname matches one of these
+    values will be ignored. :ref:`Regular expression syntax <re-syntax>` is
+    supported:
```
Please don't re-wrap this. It fits perfectly on two lines if the text is allowed to be 80 columns instead of 79.

dgw commented on sopel/config/core_section.py:
```diff
@@ -389,57 +564,86 @@ def homedir(self):
+
+    .. code-block:: ini
+
+        log_raw = yes
```
```suggestion
        log_raw = True
```

I'm partial to using the Pythonic form of a Boolean for these, but you do you. 😹

Exirel replied:
As discussed on IRC, I don't want to give the impression that it's a Python value (like it would be evaluated as such), when it's not.

I'll also add a small section to explain that you can use more than "yes", but also "y", "on", etc. for any boolean value.

dgw commented on sopel/config/core_section.py:
```diff
@@ -620,18 +856,48 @@ def homedir(self):
     It is a regular expression (so the default, ``\\.``, means commands start
     with a period), though using capturing groups will create problems.
+
+    With the default value, user will invoke commands like this::
```
```suggestion
    With the default value, users will invoke commands like this::
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -620,18 +856,48 @@ def homedir(self):
+
+    With the default value, user will invoke commands like this::
+
+        <nick> .help
```
I'm too lazy today to clone the PR and run Sphinx build, but won't the above `.. highlight:: ini` also try to act on this (and the similar one below with `?`)?

Exirel replied:
Aw. It colors it light blue. Not sure I like it this way.

dgw replied:
Perhaps this section just needs to use explicit `.. code-block::` blocks with the highlighting set for each. :/

dgw commented on sopel/config/core_section.py:
```diff
@@ -620,18 +856,48 @@ def homedir(self):
+            prefix = \\?
+            help_prefix = ?
+
+        In that example, user will invoke commands like this::
```
```suggestion
        In that example, users will invoke commands like this::
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -642,11 +908,12 @@ def homedir(self):
     If ``True``, Sopel will send information about the triggered exception to
     the sender of the message that caused the error.
 
-    If ``False``, Sopel will only log the error and will appear to fail silently
-    from the triggering IRC user's perspective.
+    If ``False``, Sopel will only log the error and will appear to fail
+    silently from the triggering IRC user's perspective.
```
Another 79 vs. 80 wrapping thing. Sigh. Unnecessary, but at least this one doesn't add any extra lines.

dgw commented on sopel/config/core_section.py:
```diff
@@ -712,24 +991,51 @@ def homedir(self):
 
     :default: ``120``
+
+    You can increase the timout (for example up to 200s) like this:
+
+    .. code-block:: ini
+
+        timeout = 200
```
```suggestion
    You can change the timeout like this:

    .. code-block:: ini

        # increase to 200 seconds
        timeout = 200
```

Not necessarily the best alternative, but I don't like the parenthetical here, and I don't like implying that the timeout can be only increased.

Exirel replied:
Good point!

dgw commented on sopel/config/core_section.py:
```diff
@@ -712,24 +991,51 @@ def homedir(self):
+
+    .. code-block:: ini
+
+        use_ssl = yes
```
```suggestion
        use_ssl = True
```

Same editorial comment as `log_raw`.

dgw commented on sopel/config/core_section.py:
```diff
@@ -712,24 +991,51 @@ def homedir(self):
+
+    .. code-block:: ini
+
+        use_ssl = yes
+        verify_ssl = yes
```
```suggestion
        use_ssl = True
        verify_ssl = True
```

Same editorial comment as `log_raw`.

Exirel commented:
I believe I addressed all the requested changes and suggestions the best I could. I've tweaked the text when they generate orphan (and I made a joke about it in the commit message because I'm feeling extra cheeky today). I also explained more about the Config Parser thing, and what choice, list, and bool options are, and how to set them properly.

I modified the prefix/help_prefix doc a bit, to make it more clear, too.

Exirel commented on sopel/config/core_section.py:
```diff
@@ -578,16 +578,19 @@ def homedir(self):
-            domain\\.com
-            .*\\.domain\\.com
+            exact-match.com
+            .*\\.regex-domain\\.com
```
There is an ``or`` though, that uses exact match. What would be the correct phrasing for that?

dgw replied:
It's not a simple matter, is it? Technically, both regex and exact-match _are_ supported. The issue is with managing the user's expectations (with a side of "why does the code try to treat the same value two ways?!", because technical debt).

The possibility for overlap between exact-match and regex syntax is probably greater for nicks than hosts (due to brackets/braces being allowed in nicknames), but it theoretically exists for both.

Maybe it would be enough to mitigate the possibility for confusion by attempting an exact match first, and only trying the regex path if that fails? Then a naïve user accidentally using regex syntax in what they think is an exact match won't be confused that their block rule didn't work.

On the other hand, trying exact matches first has its own potential for accidental false-positives on the nickname side… But users who choose regex-like nicknames deserve to be blocked anyway. 😛

Exirel replied:
I see that there is a problem, but I'm not sure what you want me to do about it in this PR.

Since I'm a bit confused, the safe path for me is to revert to a more elusive explanation for now. Something like: "Regular expression are supported (remember to escape characters as needed)."?

I could also add a little warning, something like "We are working on a better block system, see issue [link] for more information and update". I feel like this approach is honest with where the code is at the moment (as you mentioned: technical debt), and I believe honest communication will help the user understand why it isn't that simple.

What do you say?

Exirel replied:
![Screen Shot 2020-05-22 at 17 05 07](https://user-images.githubusercontent.com/3259552/82681731-7acf2780-9c4e-11ea-8f36-d283c40dd889.png)

Maybe it could be a "note" instead of a warning, but I wanted you to see how it looks like first.

dgw replied:
That screenshot looks good, though I agree about using `.. note::` instead.

I assume this is for both `host_blocks` and `nick_blocks`? Maybe the `.. warning::` would be the bit about regular expression syntax—but that might be too many blocks. 😅

Exirel replied:
Done!

Exirel commented on sopel/config/core_section.py:
```diff
@@ -872,7 +885,7 @@ def homedir(self):
 
     prefix = ValidatedAttribute('prefix', default='\\.')
-    """The prefix to add to the beginning of commands.
+    """The prefix to add to the beginning of commands as a regular expression.
```
I really hope so!

dgw commented on sopel/config/core_section.py:
```diff
@@ -813,9 +816,19 @@ def homedir(self):
     """A list of nicks which Sopel should ignore.
 
     Messages from any user whose nickname matches one of these values will be
-    ignored. :ref:`Regular expression syntax <re-syntax>` is supported.
+    ignored. Both :ref:`regular expression syntax <re-syntax>` and exact match
+    are supported:
+
+    .. code-block:: ini
+
+        nick_blocks =
+            ExactNick
+            _*RegexMatch_*
```
We'll confine discussion to just the one thread: https://github.com/sopel-irc/sopel/pull/1866#discussion_r428909631

dgw commented on sopel/config/core_section.py:
```diff
@@ -578,19 +578,25 @@ def homedir(self):
 
+    .. note::
+
+        We are working on a better block system, see `issue #1355`__ for more
```
```suggestion
        We are working on a better block system; see `issue #1355`__ for more
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -829,6 +835,13 @@ def homedir(self):
 
+    .. note::
+
+        We are working on a better block system, see `issue #1355`__ for more
```
```suggestion
        We are working on a better block system; see `issue #1355`__ for more
```

Exirel commented:
@dgw All good now. Even Travis is happy.

dgw commented on docs/source/configuration.rst:
```diff
@@ -30,14 +30,46 @@ This file can be generated with ``sopel configure``.
+Python you are running your instance of Sopel with (indentation and comments
+in particular are handled differently between Python 2.7 and Python 3+).
+
+On top of that, Sopel uses its own configuration class to handle certain type
+of options, such as choice, integer, float, list, and boolean options.
```
```suggestion
On top of that, Sopel uses its own configuration class to handle certain types
of options, such as choice, integer, float, list, and boolean.
```
